File: 代码随想录/08_回溯算法/02_组合/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int n, k;
    vector<vector<int>> res;
    vector<int> path;
    vector<vector<int>> combine(int n, int k) {
        this->n = n;
        this->k = k;
        dfs(path, 1);
        return res;
    }

    void dfs(vector<int> path, int u) {
        if (k == path.size()) {
            res.push_back(path);
            return;
        }
        // 剪枝操作
        for (int i = u; i <= n - (k - path.size()) + 1; i++) {
            path.push_back(i);
            dfs(path, i + 1);
            path.pop_back();
        }
    }
};

File: 代码随想录/08_回溯算法/02_组合/Solution.java
import java.util.ArrayList;
import java.util.List;

// https://leetcode.cn/problems/combinations/
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    int n, k;

    public List<List<Integer>> combine(int n, int k) {
        this.n = n;
        this.k = k;
        dfs(new ArrayList<>(), 1);
        return res;
    }

    void dfs(List<Integer> path, int begin) {
        if (path.size() == k) {
            res.add(new ArrayList<>(path));
            return;
        }
        // 剪枝
        for (int i = begin; i <= n - (k - path.size()) + 1; i++) {
            path.add(i);
            dfs(path, i + 1);
            path.remove(path.size() - 1); // 回溯
        }
    }
}